content/advent-2017/property-based-testing.md

```diff
@@ -0,0 +1,340 @@
+# Properties #
+
+Instead of doing that, we can adopt a different approach to testing. Instead of asking "what is the expected result of this program given these example inputs?", we ask the question: "what is the property of the result and program that doesn't change given inputs?"
+
+All that sounds a bit abstract. But applying it on the `Add` function that we've become so familiar - what doesn't change in `Add`? 
+
+Well, we know through much real life experience that `a + b` is equal to `b + a`. Mathematicians call this property [commutativity](https://en.wikipedia.org/wiki/Commutative_property). So test for it. Write a test `Add`, that takes two pairs of random inputs, perform them once, and perform them with the operands' orders switched. They should have the same result.
+
+Of course, you could have a test that reads something like this:
+
+```go
+func TestAdd(t *testing.T) {
+	// a function that checks for commutativity
+	comm := func(fn func(a, b int) int) bool {...} 
+	if !comm(Add) {
+		t.Error("Not Commutative")
+	}
+}
+
+func Add(a, b int) int { return a * b } // PEBKAC error
+```
+
+The test above will pass. But it'd be doing the wrong thing. Here you need to be asking, what makes `Add` different from `Mul`? Or more generally: What makes any function that has a signature `func(a, b int) int` different from `Add`?
+
+There are obviously a few guidelines, which I will list below, but before that, do have a think. 
+
+The point I'm trying to make here is that there are multiple properties to be tested for every program. You can't just test one property and be done with it. 
+
+# How To Do Property Based Testing In Go #
+
+So, with the Whys out of the way, let's talk about how to do property testing in Go. Go actually comes with a built in property testing tool, `"testing/quick"`. Given that property based testing is based on Haskell's QuickCheck, naturally the testing function is called `quick.Check`. The gist of how it works is simple: write your property test as a function, pass it into the `quick.Check` function, and Bob's your uncle.
+
+Here's how to test for commutativity with `"testing/quick"`:
+
+```go
+import "testing/quick"
+
+func TestAdd(t *testing.T) {
+	// property test
+	comm := func(a, b int) bool {
+		if Add(a, b) != Add(b, a) {
+			return false
+		}
+		return true
+	}
+	if err := quick.Check(comm, nil); err != nil {
+		t.Error(err)
+	}
+}
+```
+
+Here we see that `comm` was defined rather simply: `if Add(a, b) != Add(b, a)` then we say that the `Add` function does not fulfil the commutativity property.
+
+What the `quick.Check` function then does is it generates values based on the input arguments of the testing function. In this case, `comm` is a function that takes two `int`. The package understands that it needs to generate two `int` to be fed into the function. 
+
+The package works for non-primitive types too. Further extensions to functionality can be had by types that implement the `quick.Generator` interface.
```

Comment:
Give small example of something with the `Generator` interface.  Maybe integers that are only even or something?  I've also used it for to generate slices of integers with smaller abolute values to make them easier to read.

Reply:
ditto to the above

Reply:
Fixed